miasm2/core/cpu.py

```diff
@@ -361,29 +254,187 @@ def neg_int(t):
-                                    (logicop, 2, pyparsing.opAssoc.RIGHT,
-                                     parse_op),
-                                    (signop, 1, pyparsing.opAssoc.RIGHT,
-                                     parse_op),
-                                    (multop, 2, pyparsing.opAssoc.LEFT,
-                                     parse_op),
-                                    (plusop, 2, pyparsing.opAssoc.LEFT,
-                                     parse_op),
-                                    ])
-    return variable, operand, base_expr
+##########################
+
+def literal_list(l):
+    l = l[:]
+    l.sort()
+    l = l[::-1]
+    o = pyparsing.Literal(l[0])
+    for x in l[1:]:
+        o |= pyparsing.Literal(x)
+    return o
+
+
+def cb_int(t):
+    assert len(t) == 1
+    integer = AstInt(t[0])
+    return integer
+
+
+def cb_parse_id(t):
+    assert len(t) == 1
+    reg = t[0]
+    return AstId(reg)
 
 
-variable, operand, base_expr = gen_base_expr()
+def cb_reg2exprid(t):
+    assert len(t) == 1
+    reg = t[0]
+    return AstId(reg)
+
+
+def cb_deref_reg(t):
+    assert len(t) == 1
+    reg = t[0]
+    ptr = reg
+    mem = AstMem(ptr, 32)
+    return mem
+
+
+def cb_deref_reg_p_int(t):
+    assert len(t) == 2
+    reg, integer = t
+
+    ptr = AstOp("+", reg, integer)
+    mem = AstMem(ptr, 32)
+    return mem
+
+
+def cb_deref_reg_p_reg(t):
+    assert len(t) == 2
+    reg1, reg2 = t
+
+    ptr = AstOp("+", reg1, reg2)
+    mem = AstMem(ptr, 32)
+    return mem
+
+
+def cb_op_not(t):
+    tokens = t[0]
+    assert len(tokens) == 2
+    result = AstOp("!", tokens[1])
+    return result
+
+
+def merge_ops(tokens, op):
+    args = []
+    if len(tokens) >= 3:
+        args = [tokens.pop(0)]
+        i = 0
+        while i < len(tokens):
+            op_tmp = tokens[i]
+            arg = tokens[i+1]
+            i += 2
+            if op_tmp != op:
+                raise ValueError("Bad operator")
+            args.append(arg)
+    result = AstOp(op, *args)
+    return result
+
+
+def cb_op_and(t):
+    result = merge_ops(t[0], "&")
+    return result
+
+
+def cb_op_xor(t):
+    result = merge_ops(t[0], "^")
+    return result
+
+
+def cb_op_sign(t):
+    assert len(t) == 1
+    op, value = t[0]
+    return -value
+
+
+def cb_op_div(t):
+    tokens = t[0]
+    assert len(tokens) == 3
+    result = AstOp("/", tokens[0], tokens[2])
+    return result
+
+
+def cb_op_plusminus(t):
+    tokens = t[0]
+    if len(tokens) == 3:
+        # binary op
+        assert isinstance(tokens[0], AstNode)
+        assert isinstance(tokens[2], AstNode)
+        op, args = tokens[1], [tokens[0], tokens[2]]
+    elif len(tokens) > 3:
+        args = [tokens.pop(0)]
+        i = 0
+        while i < len(tokens):
+            op = tokens[i]
+            arg = tokens[i+1]
+            i += 2
+            if op == '-':
+                if isinstance(arg, AstInt):
```

Comment:
This is already done in `Ast` class code. `- arg` is then enough

Reply:
Ack

Reply:
Well, let me be more precise. From `AstNode` code:
```Python
    def __neg__(self):
        if isinstance(self, AstInt):
            value = AstInt(-self.value)
        else:
            value = AstOp('-', self)
        return value
```
So the whole `if isinstance(..., AstInt): ...; else: arg = -arg` is actually `arg = - arg`